messaging/src/main/java/org/axonframework/commandhandling/gateway/DefaultReactorCommandGateway.java

```diff
@@ -0,0 +1,253 @@
+     * </p>
+     *
+     * @return a Builder to be able to create a {@link DefaultReactorCommandGateway}
+     */
+    public static Builder builder() {
+        return new Builder();
+    }
+
+    @Override
+    public <R> Mono<R> send(Object command) {
+        //noinspection unchecked
+        return Mono.<CommandMessage<?>>just(GenericCommandMessage.asCommandMessage(command))
+                .transform(this::processCommandInterceptors)
+                .flatMap(this::dispatchCommand)
+                .flatMap(this::processResultsInterceptors)
+                .transform(this::getPayload);
+    }
+
+    @Override
+    public Registration registerDispatchInterceptor(ReactorMessageDispatchInterceptor<CommandMessage<?>> interceptor) {
+        dispatchInterceptors.add(interceptor);
+        return () -> dispatchInterceptors.remove(interceptor);
+    }
+
+    @Override
+    public Registration registerResultHandlerInterceptor(
+            ReactorResultHandlerInterceptor<CommandMessage<?>, CommandResultMessage<?>> interceptor) {
+        resultInterceptors.add(interceptor);
+        return () -> resultInterceptors.remove(interceptor);
+    }
+
+    private Mono<CommandMessage<?>> processCommandInterceptors(Mono<CommandMessage<?>> commandMessage) {
+        return Flux.fromIterable(dispatchInterceptors)
+                   .reduce(commandMessage, (command, interceptor) -> interceptor.intercept(command))
+                   .flatMap(Mono::from);
+    }
+
+    private <C, R> Mono<Tuple2<CommandMessage<C>, Flux<CommandResultMessage<? extends R>>>> dispatchCommand(
+            CommandMessage<C> commandMessage) {
+        ReactiveCallback<C, R> reactiveCallback = new ReactiveCallback<>();
+        CommandCallback<C, R> callback = reactiveCallback;
+        if (retryScheduler != null) {
+            callback = new RetryingCallback<>(callback, retryScheduler, commandBus);
+        }
+        commandBus.dispatch(commandMessage, callback);
+        return Mono.just(commandMessage).zipWith(Mono.just(Flux.from(reactiveCallback)));
+    }
+
+    private <C> Mono<? extends CommandResultMessage<?>> processResultsInterceptors(
+            Tuple2<CommandMessage<C>, Flux<CommandResultMessage<?>>> commandWithResults) {
+        CommandMessage<?> commandMessage = commandWithResults.getT1();
+        Flux<CommandResultMessage<?>> commandResultMessages = commandWithResults.getT2();
+        return Flux.fromIterable(resultInterceptors)
+                   .reduce(commandResultMessages,
+                           (result, interceptor) -> interceptor.intercept(commandMessage, result))
+                   .flatMap(Flux::next); // command handlers provide only one result!
+    }
+
+    private <R> Mono<R> getPayload(Mono<? extends CommandResultMessage<?>> commandResultMessage) {
+        //noinspection unchecked
+        return commandResultMessage
+                .filter(r -> Objects.nonNull(r.getPayload()))
+                .map(it -> (R) it.getPayload());
```

Comment:
The reactor doesn't support `null` values in a stream. Null value is equivalent of` Mono.empty()` in this case. Filter operator archive's that.

Reply:
What does it mean to be archived? Does that mean that users can still see there were `null` responses?

I'd argue that `null` is a valid response for some queries, hence why I am asking for it.
You could however further argue that if users are interested in this, that they can use the `OptionalResponseType` to better tie into that.

Anyhow, what I am angling for, is there a way to get such a result to users at all?

Reply:
sorry not `achieve`, its `achieve`. so if query returns null it will be mapped to Mono.empty, and user can react on that as it can on `null` value, but in a reactive, functional way. Instead of using optional or writing if statements, different operators like `.switchIfEmpty` could be used to handle `null` return values. Additionally, as you mentioned user can introduce `OptionalResponseType`, but I see that as an anti-pattern, since Mono offers the same functionality (Optional.empty = Mono.empty)

Reply:
Gotcha, thanks for explaining this further @stefanvozd!